python/dgl/graphbolt/graph_storage/csc_sampling_graph.py

```diff
@@ -0,0 +1,267 @@
+"""CSC format sampling graph."""
+# pylint: disable= invalid-name
+from typing import Dict, Optional, Tuple
+
+import torch
+
+
+class GraphMetadata:
+    r"""Class for metadata of csc sampling graph."""
+
+    def __init__(
+        self,
+        node_type_to_id: Dict[str, int],
+        edge_type_to_id: Dict[Tuple[str, str, str], int],
+    ):
+        """
+        Initialize the GraphMetadata object.
+
+        Parameters
+        ----------
+        node_type_to_id : Dict[str, int]
+            Dictionary from node types to node IDs.
+        edge_type_to_id : Dict[Tuple[str, str, str], int]
+            Dictionary from edge types to edge IDs.
+
+        Raises
+        ------
+        AssertionError
+            If any of the assertions fail.
+        """
+
+        node_types = list(node_type_to_id.keys())
+        edge_types = list(edge_type_to_id.keys())
+        node_type_ids = list(node_type_to_id.values())
+        edge_type_ids = list(edge_type_to_id.values())
+
+        assert all(
+            isinstance(x, str) for x in node_types
+        ), "Node type name should be string."
+        assert all(
+            isinstance(x, int) for x in node_type_ids
+        ), "Node type id should be int."
+        assert all(
+            isinstance(x, int) for x in edge_type_ids
+        ), "Edge type id should be int."
+        assert len(node_type_ids) == len(
+            set(node_type_ids)
+        ), "Multiple node types shoud not be mapped to a same id."
+        assert len(edge_type_ids) == len(
+            set(edge_type_ids)
+        ), "Multiple edge types shoud not be mapped to a same id."
+        edges = set()
+        for edge_type in edge_types:
+            src, edge, dst = edge_type
+            assert isinstance(edge, str), "Edge type name should be string."
+            assert edge not in edges, f"Edge type {edge} is defined repeatedly."
+            edges.add(edge)
+            assert (
+                src in node_types
+            ), f"Unrecognized node type {src} in edge type {edge_type}"
+            assert (
+                dst in node_types
+            ), f"Unrecognized node type {dst} in edge type {edge_type}"
+        self.node_type_to_id = node_type_to_id
+        self.edge_type_to_id = edge_type_to_id
+
+
+class CSCSamplingGraph:
+    r"""Class for CSC sampling graph."""
+
+    def __repr__(self):
+        return _csc_sampling_graph_str(self)
+
+    def __init__(
+        self, c_csc_graph: torch.ScriptObject, metadata: Optional[GraphMetadata]
+    ):
+        self.c_csc_graph = c_csc_graph
+        self.metadata = metadata
```

Comment:
I wonder when will `metadata` be None? Technically even for homogeneous graph, there is one node type and one edge type.

Reply:
let's keep it optional for homogeneous graph, usually people don't think about edge type / node type when they play with homogeneous graph, make it required will lead to some confusion.